src/analysis/calculus/extend_deriv.lean

```diff
@@ -0,0 +1,217 @@
+  have cxnt : cx n ∈ t := ⟨cxs n, hnx⟩,
+  have cynt : cy n ∈ t,
+  { refine ⟨cys n, _⟩,
+    calc dist (cy n) x ≤ dist (cy n) y + dist y x : dist_triangle _ _ _
+    ... < δ/2 + δ/2 : add_lt_add hny hy.1
+    ... = δ : by ring },
+  refine norm_image_sub_le_of_norm_deriv_le_convex diff_t (λz zt, _) t_conv cxnt cynt,
+  have : fderiv_within ℝ f t z = fderiv ℝ f z,
+  { have t_open : is_open t := is_open_inter s_open is_open_ball,
+    rw differentiable_at.fderiv_within _ (t_open.unique_diff_on z zt),
+    apply (diff_t z zt).differentiable_at,
+    apply mem_nhds_sets t_open zt },
+  rw [this, ← dist_zero_right],
+  exact le_of_lt (hδ zt.1 zt.2)
+end
+
+/-- If a function `f` is differentiable in a convex open set and continuous on its closure, and its
+derivative converges to a limit `f'` at a point on the boundary, then `f` is differentiable there
+with derivative `f'`. -/
+theorem has_fderiv_at_boundary_of_tendsto_fderiv {f : E → F} {s : set E} {x : E} {f' : E →L[ℝ] F}
+  (f_diff : differentiable_on ℝ f s) (s_conv : convex s) (s_open : is_open s)
+  (f_cont : ∀y ∈ closure s, continuous_within_at f s y)
+  (h : tendsto (λy, fderiv ℝ f y) (nhds_within x s) (𝓝 f')) :
+  has_fderiv_within_at f f' (closure s) x :=
+begin
+  /- We subtract `f'` to define a new function `g` for which `g' = 0`, for which differentiability
+  is proved `has_fderiv_at_boundary_of_differentiable_aux`. Then, we just need to glue together the
+  pieces, expressing back `f` in terms of `g`. -/
+  let g := λy, f y - f' y,
+  have diff_g : differentiable_on ℝ g s :=
+    f_diff.sub (f'.differentiable.comp differentiable_id).differentiable_on,
+  have cont_g : ∀y ∈ closure s, continuous_within_at g s y :=
+    λy hy, tendsto.sub (f_cont y hy) (f'.continuous.comp continuous_id).continuous_within_at,
+  have A : ∀y ∈ s, fderiv ℝ f y - f' = fderiv ℝ g y,
+  { assume y hy,
+    have : has_fderiv_at f (fderiv ℝ f y) y :=
+      (differentiable_within_at.differentiable_at (f_diff y hy) (mem_nhds_sets s_open hy)).has_fderiv_at,
+    have : has_fderiv_at g (fderiv ℝ f y - f') y :=
+      this.sub (f'.has_fderiv_at.comp y (has_fderiv_at_id y)),
+    exact this.fderiv.symm },
+  have B : tendsto (λy, fderiv ℝ f y - f') (nhds_within x s) (𝓝 (f' - f')) :=
+    h.sub tendsto_const_nhds,
+  have : tendsto (λy, fderiv ℝ g y) (nhds_within x s) (𝓝 0),
+  { have : f' - f' = 0, by simp,
+    rw this at B,
+    apply tendsto.congr' _ B,
+    filter_upwards [self_mem_nhds_within] A },
+  have : has_fderiv_within_at g (0 : E →L[ℝ] F) (closure s) x :=
+    has_fderiv_at_boundary_of_tendsto_fderiv_aux diff_g s_conv s_open cont_g this,
+  convert this.add f'.has_fderiv_within_at,
+  { ext y, simp [g] },
+  { simp }
+end
+
+/-- If a function is differentiable on the right of a point `a : ℝ`, continuous at `a`, and
+its derivative also converges at `a`, then `f` is differentiable on the right at `a`. -/
+lemma has_deriv_at_interval_left_endpoint_of_tendsto_deriv {s : set ℝ} {e : E} {a : ℝ} {f : ℝ → E}
```

Comment:
I find it unfortunate that these specialisation lemmas still need proofs of >20 lines. Does this mean that some API parts are missing, or maybe some simp-lemmas somewhere? Especially, since for the "dual" version you need essentially the same proof, which is another >20 lines.

Reply:
The thing is that the main theorem is rather complicated, with 5 different assumptions. If you want to use the theorem in a different context (one-dimensional), all these assumptions come under a different formulation, so there is necessarily a translation between the two contexts. It is mainly straightforward, so I don't think it is that bad. I could certainly divide the length of the proof by a factor of 2, at the cost of readability.

For the dual version, I could also have gone by composing with `x -> -x`, applying the direct version, and then composing with `x -> -x` again. But I doubt it would be shorter (you would need to play with composition of derivatives, check where domains are sent, and so on), so by lazyness I copy-pasted the proof.